_posts/2021/2021-03-31-hmat_memkind.md

```diff
@@ -0,0 +1,172 @@
+---
+title: Memkind support for heterogeneous memory attributes
+author: michalbiesek
+layout: post
+identifier: hmat_memkind
+---
+
+### Introduction
+
+Memkind is a library that provides support for Persistent Memory. However, it is not the only type of memory that memkind supports.
+It is a general solution designated for the platforms with heterogeneous memory.
+
+But before we go to heterogenous memory itself, let's start with a short recap about [NUMA][kernel-numa].
+The NUMA concept aims the problem of the dynamic extension of CPU count per socket and system memory.
+Before NUMA, commonly used model was uniform memory access (UMA), where all processors share the physical memory uniformly.
+With the UMA approach and rapid growth of number of processors and memory in multi-socket machines, platforms now face scalability problems.
+Single shared bus between all CPU(s) and DRAM memory become a bottleneck, as each CPU decreases the available bandwidth.
+To handle each new device, the length of the bus must be extended, what in turn impacts the latency.
+
+To overcome the challenges mentioned above, engineers decide to introduce a new strategy: None Unified Memory Access(NUMA).
+Instead of providing a single shared bus, the concept proposes a solution to divide the hardware resources into `NUMA Nodes` abstraction.
+Depending on the exact implementation on multi-socket platforms, NUMA Node could be exposed as CPU(s) and memory placed inside the same socket.
+Accessing the memory placed on the same socket as the cpu on which process runs (`local memory access`) is faster since it avoids
+additional overhead from interconnection bus, like the memory access between two sockets (`remote memory access`).
+
+<figure class="image">
+  <img src="/assets/numa_overview.png" alt="NUMA overview">
+  <figcaption>Figure 1. NUMA overview.</figcaption>
+</figure>
+
+The operating system made the memory allocation decision among other factors based on `NUMA distance.` The distance represents
+relative memory latency between NUMA nodes, and is based on hardware configuration - System Locality Information Table (SLIT) and
+System Resource Affinity Table (SRAT) presented in [ACPI][ACPI-doc].
+NUMA aware system Operating System can still consume all possible memory, but the OS will optimize it for local memory access.
+In practice, this means that memory allocation will come from the NUMA Node with the smallest distance from the CPU of the calling thread.
+
+NUMA memory access pattern could be decribed as **heterogenous memory access**.
+
+But today's NUMA Node concept is vague compared to the original one. Extension of different Processing Units
+provides **heterogenous computing**, but the memory doesn't stay behind in this race. The NUMA Node itself now
+doesn't have to be a couple between DRAM and CPU.
```

Comment:
don't have to contain both memory (typically DRAM) and compute (typically a CPU)